packages/perspective-viewer-d3fc/src/js/tooltip/tooltip.js

```diff
@@ -0,0 +1,92 @@
+import {select} from "d3";
+import {getChartElement} from "../plugin/root";
+import tooltipTemplate from "../../html/tooltip.html";
+
+export function tooltip(selection, settings) {
+    const container = select(getChartElement(selection.node()).getContainer());
+    const tooltipDiv = getTooltipDiv(container);
+    selection
+        .filter(d => d.baseValue !== d.mainValue)
+        .on("mouseover", function(data) {
+            generateHtml(tooltipDiv, data, settings);
+            showTooltip(container.node(), this, tooltipDiv);
+        })
+        .on("mouseout", function() {
+            hideTooltip(tooltipDiv);
+        });
+}
+
+function getTooltipDiv(container) {
+    let tooltipDiv = container.select("div.tooltip");
+    if (tooltipDiv.size() === 0) {
+        tooltipDiv = container
+            .append("div")
+            .attr("class", "tooltip")
+            .style("z-index", 3)
+            .style("opacity", 0)
+            .html(tooltipTemplate);
+    }
+    return tooltipDiv;
+}
+
+function generateHtml(tooltipDiv, data, settings) {
+    const splits = data.key.split("|");
+
+    // Add group data
+    if (settings.crossValues.length) {
+        const groups = data.crossValue.split(",");
+        tooltipDiv
+            .select("#cross-values")
+            .selectAll("li")
+            .data(groups)
+            .join("li")
+            .each(function(d, i) {
+                select(this)
```

Comment:
Now you could pull out the function like this:

````javascript
function eachValue(list) {
    (d, i) => {
        select(this)
            .text(`${list[i].name}: `)
            .append("b")
            .text(d);
    }
}
````

and 

````javascript
    .data(groups)
            .join("li")
            .each(eachValue(settings.crossValues));
````

Reply:
Sounds good